crates/bdk/tests/wallet.rs

```diff
@@ -60,6 +60,21 @@ fn receive_output_in_latest_block(wallet: &mut Wallet, value: u64) -> OutPoint {
 // OP_PUSH.
 const P2WPKH_FAKE_WITNESS_SIZE: usize = 106;
 
+#[test]
+fn test_which_keychain_derived_script() {
+    let (mut segwit_wallet, _) = get_funded_wallet(get_test_wpkh());
+    let addr = segwit_wallet.get_address(New);
+    let script = addr.script_pubkey();
+    let keychain = segwit_wallet.which_keychain_derived(&script).unwrap();
+    assert_eq!(keychain.0, KeychainKind::External);
+
+    let (mut taproot_wallet, _) = get_funded_wallet(get_test_tr_single_sig());
+    let addr = taproot_wallet.get_address(New);
+    let script = addr.script_pubkey();
+    let keychain = segwit_wallet.which_keychain_derived(&script);
+    assert!(keychain.is_none());
```

Comment:
In addition to checking the keychain, we could possibly also check the correct spk index is returned from `which_keychain_derived`.

<details>

<summary>diff</summary>

```diff

 fn test_which_keychain_derived_script() {
     let (mut segwit_wallet, _) = get_funded_wallet(get_test_wpkh());
     let addr = segwit_wallet.get_address(New);
     let script = addr.script_pubkey();
     let keychain = segwit_wallet.which_keychain_derived(&script).unwrap();
     assert_eq!(keychain.0, KeychainKind::External);
 
     let (mut taproot_wallet, _) = get_funded_wallet(get_test_tr_single_sig());
     let addr = taproot_wallet.get_address(New);
     let script = addr.script_pubkey();
     let keychain = segwit_wallet.which_keychain_derived(&script);
     assert!(keychain.is_none());
+
+    // find correct derivation index from spk
+    let desc = get_test_tr_single_sig_xprv();
+    let mut wallet = Wallet::new_no_persist(desc, None, Network::Testnet).unwrap();
+    let spk = wallet.get_address(New).address.script_pubkey();
+    let (_k, i) = wallet.which_keychain_derived(&spk).unwrap();
+    assert_eq!(i, &0);
+
+    // reveal more addresses. the last revealed index should now be 10.
+    for _ in 0..10 {
+        let _ = wallet.get_address(New);
+    }
+    let spk = wallet.get_address(Peek(10)).address.script_pubkey();
+    let (_k, i) = wallet.which_keychain_derived(&spk).unwrap();
+    assert_eq!(i, &10);
 }

```

</details>